plugins/dell-dock/fu-plugin-dell-dock.c

```diff
@@ -43,6 +49,7 @@ fu_plugin_init (FuPlugin *plugin)
 
 	/* currently slower performance, but more reliable in corner cases */
 	fu_plugin_add_rule (plugin, FU_PLUGIN_RULE_BETTER_THAN, "synaptics_mst");
+	fu_plugin_add_rule (plugin, FU_PLUGIN_RULE_BETTER_THAN, "thunderbolt");
```

Comment:
@superm1 this seems never proven to work. 
```

18:13:20:0575 FuPluginList         uefi_capsule [0] to be ordered after upower [0] so promoting to [1]
18:13:20:0575 FuPluginList         dell [0] to be ordered after uefi_capsule [1] so promoting to [2]
18:13:20:0575 FuPluginList         uefi_recovery [0] to be ordered after uefi_capsule [1] so promoting to [2]
18:13:20:0575 FuPluginList         tpm_eventlog [0] to be ordered after tpm [0] so promoting to [1]
18:13:20:0575 FuPluginList         tpm_eventlog [1] to be ordered before uefi_capsule [1] so promoting to [2]
18:13:20:0575 FuPluginList         dell [2] to be ordered after uefi_capsule [2] so promoting to [3]
18:13:20:0575 FuPluginList         uefi_recovery [2] to be ordered after uefi_capsule [2] so promoting to [3]
18:13:20:0575 FuPluginList         cpu [0] to be ordered before msr [0] so promoting to [1]
18:13:20:0575 FuPluginList         dell_esrt [0] better than bios [0] so bumping to [1]
18:13:20:0575 FuPluginList         dell_dock [0] better than synaptics_mst [0] so bumping to [1]
18:13:20:0575 FuPluginList         dell [0] better than tpm [0] so bumping to [1]
18:13:20:0575 FuPluginList         bcm57xx [0] better than optionrom [0] so bumping to [1]

```

Reply:
Can we set priority on device level or masking another device? I thought priority only available in plugin level.  

I tried to revert  the changes by 078beaf but lots errors are waiting to be fixed.

Reply:
Yeah I don't think revert the whole commit, just the logic in `src/fu-device-list.c`.  Hopefully that's not still painful to revert.

@hughsie I think setting plugin priority implicitly sets device priority, but the key here is that after 078beaf the daemon won't do anything about two plugins that provide the same device (which is exactly the case here - `thunderbolt` plugin and `dell_dock` plugin both provide a representation of the USB4 controller), but the goal is to have `dell_dock` plugin (or `intel-goshen-ridge` if it splits out) take precedence.

The reason for that preference is that it removes the kernel and the type of port you're connected to from the equation when it comes to the firmware update.  Otherwise you'll need 5.13 or commits backported to handle USB4 updates on this module,

Reply:
> the daemon won't do anything about two plugins that provide the same device

I don't think we want to go back to the device list deduping things like we used to do. Can we achieve the same thing by ordering dell-dock before thunderbolt and then handling the ->registered vfunc? i.e. register dell_dock, then if the thunderbolt dock shows up then just fu_device_inhibit() it.

Reply:
I guess we need to think about whether `FU_PLUGIN_RULE_BETTER_THAN` really makes sense anymore.  Did we unintentionally for example break the case of dell plugin tpm vs tpm plugin?  Or dell-esrt vs bios plugin?

>I don't think we want to go back to the device list deduping things like we used to do. Can we achieve the same thing by ordering dell-dock before thunderbolt and then handling the ->registered vfunc? i.e. register dell_dock, then if the thunderbolt dock shows up then just fu_device_inhibit() it.

I'm worried about circular dependencies with this design pattern.  I think if we're going to do inhibitions, they should work no matter what order the plugins load.  So that means that `dell_dock` or `intel-goshen-ridge` would need to look for a device coming from `thunderbolt` and determine whether it's something it can also handle, maybe via a FuDevice metadata.  The question then becomes how to tell when to give that metadata to the device?  Probably need to parse NVM in `fu-thunderbolt-firmware.c` to tell it's GR and then apply some metadata.

Reply:
Lets go back a bit; `FU_PLUGIN_RULE_BETTER_THAN` allows the plugin list to set the _plugin_ priority. This allows fu_engine_plugin_device_added_cb() to set the *device* priority when the plugin emits the device added signal. This is where the plan goes horribly wrong, as the device priority doesn't seem to be used for anything any more.

Lets fix this, and add some tests. I think it will be easy to fix, **IF** the dell_dock and thunderbolt plugins produce the same device (with a different priority) with exactly the same `device_id`.

Reply:
> Lets fix this, and add some tests. I think it will be easy to fix, IF the dell_dock and thunderbolt plugins produce the same device (with a different priority) with exactly the same device_id.

Hmm, so same physical and logical id both..  I think that's going to be challenging.  The thunderbolt device will get physical ID set as `thunderbolt_device`.  This can be hardcoded in dell-dock or intel-goshen-ridge plugin, but same logical id, I'm not sure. @CragW  what logical ID is set right now for thunderbolt plugin?  Is it NULL?  If so, maybe you can hardcode `thunderbolt_device`  as physical ID now here.

Reply:
I found a system of mine on hand that had TBT controller, I don't think it will be viable to have same physical-id very likely.

```
Thunderbolt host controller
  DeviceId:             f78dd6bb609522e31f1ff2ff819827108606d18a
  Guid:                 83f70f65-8de5-5520-93e2-a569615030bb <- THUNDERBOLT\VEN_00D4&DEV_0962&REV_00
  Guid:                 b067bd0b-045e-5b00-9948-f7c98fef1f72 <- THUNDERBOLT\VEN_00D4&DEV_0962
  Guid:                 7d538854-204d-51b2-8f9d-1fe881c70200 <- TBT-00d40962-native
  Guid:                 8a1cafa7-2d8b-5641-ba4a-140a6497a5e9 <- TBT-00d40962-native-controller0-0
  Summary:              Unmatched performance for high-speed I/O
  Plugin:               thunderbolt
  Protocol:             com.intel.thunderbolt
  Flags:                internal|updatable|require-ac|registered|dual-image
  Vendor:               Dell
  VendorId:             THUNDERBOLT:0x00D4
  VendorId:             TBT:0x00D4
  Version:              44.02
  VersionFormat:        pair
  Icon:                 thunderbolt
  Created:              2021-06-11
  PhysicalId:           cc030000-0080-7718-23c8-fd1cc6132024
  BackendId:            /sys/devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:00.0/0000:05:00.0/domain0/0-0
  PossiblePlugin:       thunderbolt
  InternalFlags:        replug-match-guid
  SysfsPath:            /sys/devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:00.0/0000:05:00.0/domain0/0-0
  Subsystem:            thunderbolt
  Device Type:          Thunderbolt host controller
  Safe Mode:            false
  Native mode:          true
  Generation:           3
  AuthAttribute:        nvm_authenticate
```

Reply:
So we do have to dedupe by GUID then...?

Reply:
Yes, `NULL` as `logical ID` is unset from `thunderbolt` plugin for now so it's `0x0`.

Reply:
I have no idea how to hardcode thunderbolt's `physical ID` here, this is `USB` interface hence thunderbolt device might be unavailable? https://github.com/fwupd/fwupd/blob/master/plugins/thunderbolt/fu-thunderbolt-device.c#L280

Reply:
I think you would need to read UUID from controller - that's how kernel builds it.  This might be achievable through USB3 MMIO interface, I don't know.  This function in kernel gets it from usb4 router: `usb4_switch_read_uid`

Reply:
That seems challenging to get thunderbolt device's physical ID in dell-dock's init.. What if a standalone USB device in `dell-dock` quirk other than reusing `TBT` one, so we can make it clearly in USB, and find out the way to **mask/disable** the `TBT` device in `fu_plugin_backend_device_added` when USB device is `added`.

Reply:
OK new idea - what if you just set plugin key in `dell-dock.quirk` to `dell_dock` and then make thunderbolt plugin examine this key if present during probe/startup.  It will see it doesn't match thunderbolt and then it shouldn't probe there at all, no deduplication needed.

Reply:
That seems muted `thunderbolt` plugin entirely because `Plugin = dell_dock `key presents in `dell-dock.quirk` all the time.  I found another way it's working nevertheless it's a hardcode in the `thunderbolt` plugin too.
```
void
fu_plugin_device_added (FuPlugin *plugin, FuDevice *device)
{
	if (fu_device_has_guid(device, "TBT-00d4b071"))
		fu_plugin_device_remove (plugin, device);
}

```

Reply:
Consideration: maybe move out `TBT-00d4b071` from hardcode to `config` file as `denylist=TBT-00d4b071` to prevent `thunderbolt` plugin probing this device.

Reply:
Take a look  at this branch.
https://github.com/fwupd/fwupd/pull/new/wip/superm1/block-tbt

Does this not work?

Reply:
The `tmp` i.e. `plugin` for the `USB4 device` is `NULL` at that time, 
![Screenshot from 2021-06-13 07-52-36](https://user-images.githubusercontent.com/30334596/121791541-607f3780-cc1d-11eb-9296-90c14102ef36.png)

Reply:
For hardcode to ignore adding `TBT-00d4b071` device, I just found another way to do this [silently](https://github.com/fwupd/fwupd/blob/master/src/fu-engine.c#L6092) (without error prints), at last better than `add` and `remove`.
```
gboolean
fu_plugin_backend_device_added (FuPlugin *plugin, FuDevice *device, GError **error)
{
	guint16 vid = fu_udev_device_get_vendor (FU_UDEV_DEVICE (device));
	guint16 pid = fu_udev_device_get_model (FU_UDEV_DEVICE (device));

	if ((vid = 0xd4) && (pid == 0xb071)) {
		g_set_error (error,
			     FWUPD_ERROR,
			     FWUPD_ERROR_NOT_SUPPORTED,
			     "device intended to be handled by plugin");
		return FALSE;
	}
	return TRUE;
}

```

Reply:
Back a bit; we're talking about a tbt device that only appears when the dock is plugged in? In which case, could we add an inhibit in a quirk file for the tbt device?

Reply:
The dock has a TBT controller can be updated thru thunderbolt interface also available thru USB. More preferred the USB for comprehensive support e.g.: host without TBT. When dock is connected, and USB device is `registered` by `dell-dock`, don't want the TBT device because firmware will be updated twice.

Reply:
What Richard means is to run `fu_device_inhibit` in the dell-dock registered callback.  You can check that:
* the plugin of the device matched from thunderbolt
* the device has the matching GUID

And then call `fu_device_inhibit` on it.

Reply:
I think that's a much better way to do this kind of thing. If you need a code example let me know.

Reply:
`fu_device_inhibit` working perfectly, thanks @superm1 and @hughsie